### PR TITLE
Merge instrument cache to inserter

### LINKED ISSUE #3720: Meter provider don't work with multiple readers
### Description

I upgrade metric SDK from v0.32.0 to v0.36.0, find the MeterProvider doesn't support multiple  readers register, it only push data to frist one.

### Environment

- OS: [linux]
- Architecture: [x86]
- Go Version: [go1.18.6]
- opentelemetry-go version: [v0.36.0]

### Steps To Reproduce

1. add stdout exporter to [example/prometheus/main.go#L50](https://github.com/open-telemetry/opentelemetry-go/blob/main/example/prometheus/main.go#L50), code like:

```
stdExporter, _ := stdoutmetric.New()
exporter, _ := prometheus.New()
provider := metric.NewMeterProvider(
  metric.WithReader(
	  metric.NewPeriodicReader(stdExporter, metric.WithInterval(8*time.Second)),
  ),
  metric.WithReader(exporter),
)
```

ps: stdout metric exporter is the first reader.

2. after running example code, we will find only stdout logs output the metrics, `curl localhost:2223/metrics | grep baz_bucket` is empty.

### Expected behavior

Stdout and Prometheus http response both include the metrics.

### PR DESCRIPTION
**Pull request by MrAlias**
The current `pipeline` resolution path will only add the resolved aggregator to the pipeline when it creates one (cache miss). It will not add it if there is a cache hit. This means (since we cache instrument at the meter level, not the pipeline level) the first reader in a multiple-reader setup is the only one that will collect data for that aggregator. All other readers will have a cache hit and nothing is added to the pipeline. This is causing #3720.

This resolves #3720 by moving the instrument caching into the inserter. This means aggregators are cached at the reader level, not the meter.

### CODE REVIEW

File: sdk/metric/pipeline.go [hunk 1]
```diff
@@ -261,6 +281,12 @@ func (i *inserter[N]) Instrument(inst Instrument) ([]internal.Aggregator[N], err
 	return aggs, errs.errorOrNil()
 }
 
+// aggCV is the cached value in an aggregators cache.
+type aggCV[N int64 | float64] struct {
```
Comment (dashpole) [inline sdk/metric/pipeline.go:285]: non-blocking: I found `aggVal` to be easier to understand than `aggCV`.

**Review by dashpole — APPROVED** [review]
IIUC, we currently have something like:

Meter > Reader (pipeline) > Instrument > Aggregation

The downside seems to be that two readers that want identical data will duplicate all of the aggregations.  I wonder if, in the future it would a good idea to switch to:

Meter > Instrument > Aggregation > Reader (pipeline)

So that identical aggregations would be shared by readers.  But I'm not entirely sure that will work, and suspect it would require a large refactor to implement.

**Comment by MrAlias** [comment]
> The downside seems to be that two readers that want identical data will duplicate all of the aggregations. I wonder if, in the future it would a good idea to switch ...
> 
> So that identical aggregations would be shared by readers. But I'm not entirely sure that will work, and suspect it would require a large refactor to implement.

True the data is going to be duplicated in memory. We did this because each reader needs to have an unaffected aggregation, when one reader collects it cannot affect the data another reader would collect. Also, when we made this choice views were tied to the reader not the meter provider.

There does seem to be an optimization for when multiple readers use cumulative temporarily and the same aggregation selectors, but state would still need to be held if delta temporality is used by a reader.

Given how long it took for https://github.com/open-telemetry/opentelemetry-go/issues/3720 to be reported, I don't expect there to be many situations using multiple readers. Which makes me think that optimization isn't too high of a priority.

**Review by hanyuancheung — APPROVED** [review]

**Comment by liufuyang** [comment]
So maybe can we merge this soon? Looking forward to the next release which brings a fix of the memory issue :)

**Pull request merged**